DEPRECATIONS.md

```diff
@@ -0,0 +1,67 @@
+# Deprecations & Breaking Changes in KEDA
+
+This document describes how KEDA handles deprecations and breaking changes and should provide clarity to KEDA end-users on what to expect when upgrading to a new version of KEDA.
+
+If you are interested in the current open deprecations, we recommend checking our [GitHub Discussions](https://github.com/kedacore/keda/discussions/categories/deprecations) or the open breaking changes](https://github.com/kedacore/keda/issues?q=is%3Aopen+label%3Abreaking-change+sort%3Aupdated-desc).
+
+## Our versioning strategy
+
+KEDA follows the [Semantic Versioning](https://semver.org/) strategy and can only introduce breaking changes in new major versions.
+
+However, depending on the nature of the change, we may introduce deprecations in existing minor versions. These changes always must be backwards-compatible.
+
+Depending on the area of the change, we might not need to wait for the next major version of KEDA to introduce a breaking change.
+
+### KEDA runtime
+
+The KEDA runtime, which consists of the operator and metric server, is not allowed to make breaking changes unless the major version is incremented.
+
+There are no exceptions to this rule.
+
+### Custom Resource Definitions (CRDs)
+
+Our customer resource definitions (CRDs) have a seperate versioning from the KEDA runtime given they have an `apiVersion`.
+
+KEDA is allowed to make breaking changes to the CRDs, when a new `apiVersion` is introduced.
+
+However, the KEDA runtime must support all current API versions until the next major version of KEDA.
```

Comment:
At this moment, CRDs are in v1alpha1, I'm not saying that we shouldn't avoid breaking changes, but based on that version, we could introduce any breaking change that we want, including any element inside it like triggers or metadata.
If we are going to block breaking changes, I'd promote CRDs to v1 directly

Reply:
In the same way, I wouldn't support multiple versions till we graduate the api version, alpha or beta are literally versions which allow changes

Reply:
> At this moment, CRDs are in v1alpha1, I'm not saying that we shouldn't avoid breaking changes, but based on that version, we could introduce any breaking change that we want, including any element inside it like triggers or metadata.

It's not because you can, that you should. That would be a bad customer experience in my opinion.

I think the key take aware with our current situation and this doc is:
1. We should go to v1 soon
2. We should only support 1 stable and one preview API version for our CRD

Reply:
Once a new stable or preview version is introduced; we maintain it for X versions (ie 2) and kill it afterwards.

Reply:
Okey but, let's say that we release `scaledobjects.keda.sh/v1` in v2.9.0, and after several releases, e.g in KEDA v2.16 we release `scaledobjects.keda.sh/v2`. Based on this, we will automatically drop the support to `scaledobjects.keda.sh/v1` but that's a breaking change and we can't do it till KEDA v3.

If we try to not doing a v3 release, I think that we should be able to introduce breaking changes with minor versions. In fact, Kubernetes does it, atm k8s is in major v1 and  they introduce breaking changes with minor versions. Personally, I'd go closer with k8s feature lifecycle than semver, because we are creating a software around k8s. 

Other option could be doing a major release per year, for instance (but I think we don't want that).

Reply:
> Okey but, let's say that we release `scaledobjects.keda.sh/v1` in v2.9.0, and after several releases, e.g in KEDA v2.16 we release `scaledobjects.keda.sh/v2`. Based on this, we will automatically drop the support to `scaledobjects.keda.sh/v1` but that's a breaking change and we can't do it till KEDA v3.

Wrong, that's why the doc explicitly calls out API version and the fact that they have a separate lifecycle from the KEDA runtime so that we are allowed to do this. And for a reason, because they are completely different.

See proposal:

> Depending on the area of the change, we might not need to wait for the next major version of KEDA to introduce a breaking change.
> [...]
> Our custom resource definitions (CRDs) have a seperate versioning from the KEDA runtime given they have an `apiVersion`.
> KEDA is allowed to make breaking changes to the CRDs, when a new `apiVersion` is introduced.
>  
> However, the KEDA runtime must support all current API versions until the next major version of KEDA.

So if `scaledobjects.keda.sh/v2` is released, it is going to deprecate/remove (to define) `scaledobjects.keda.sh/v1`. I'd argue that `scaledobjects.keda.sh/v2alpha` or `scaledobjects.keda.sh/v2beta` supersedes `scaledobjects.keda.sh/v2` but the policy needs to be more explicit on this.

My proposal would be:
- A new stable CRD must have a preview CRD that supersedes it
- Once new stable CRD is introduced, the previous stable CRD version is only supported for 2 versions (worst case: 1) so that end-users can migrate
- We only support one preview API version

> Other option could be doing a major release per year, for instance (but I think we don't want that).

This is something I want to avoid.

Does this clarify things @JorTurFer @zroubalik ?

Reply:
Updated after our discussion